### PR TITLE
feat: RFC116 & RFC135

### PR DESCRIPTION
**Pull request by garylin2099**
feat: [RFC116](https://deepwisdom.feishu.cn/wiki/LmurwttAJiB8W9k1fy5cfxZMnEe)
feat: [RFC135](https://deepwisdom.feishu.cn/wiki/B1bcwZRdPiC8OYkweZYc9qWinSb)

1. Switched from shared memory to private memory, coupled with a message publishing-subscription mechanism to support distributed deployment.
2. Added Git repository and file repository utilities.
3. The `SoftwareCompany` enhanced incremental iteration capability for meeting requirements, managing versions through Git.
4. `ProductManager` introduced the `PrepareDocuments` action to create project environments like Git repositories.
5. Engineer added the `SummarizeCode` action to inspect the code from a global perspective. Its execution can be limited by the `--max-auto-summarize-code` parameter.
6. Introduced CLI mode.
7. Encapsulated the globally unique Config object into a session-level ContextVar to support the creation of multiple logically isolated instances within the same process.
8. Bugfix: Addressed the issue in the QA phase where dependencies were not automatically installed in the `RunCode` action.

中：
1. 共享memory改为私有memory，搭配消息发布订阅机制，以支持分布式部署。
2. 增加git repository和file repository。
3. `SoftwareCompany`增加了需求的增量迭代能力，通过git管理版本。
4. `ProductManager`新增`PrepareDocuments` action，用来创建git repository等项目环境；
5. `Engineer`新增`SummarizeCode` action，从全局视角检查代码。可通过--max-auto-summarize-code参数限制其执行次数。
6. 新增cli模式。
7. 将全局唯一的Config对象封装成会话粒度的ContextVar, 以支持同一进程内创建多个逻辑隔离的实例。
8. bugfix: QA阶段`RunCode` action没有自动安装依赖包的问题。

### CODE REVIEW

File: metagpt/roles/role.py [hunk 1]
```diff
@@ -291,43 +371,56 @@ async def react(self) -> Message:
...
+
+        if not await self._observe():
             # If there is no new information, suspend and wait
             logger.debug(f"{self._setting}: no news. waiting.")
             return
 
-        rsp = await self.react()
-        # Publish the reply to the environment, waiting for the next subscriber to process
-        self._publish_message(rsp)
+        rsp = await self._react()
```
Comment (garylin2099) [inline metagpt/roles/role.py:415]: 这个应该用react()，不是_react()

Comment (iorisa) [reply metagpt/roles/role.py]: fixed

File: metagpt/document.py [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+from langchain.text_splitter import CharacterTextSplitter
+from pydantic import BaseModel, Field
+from tqdm import tqdm
+
+from metagpt.config import CONFIG
+from metagpt.logs import logger
+from metagpt.repo_parser import RepoParser
+
+
+def validate_cols(content_col: str, df: pd.DataFrame):
```
Comment (better629) [inline metagpt/document.py:27]: most content repeated with `metagpt/document_store/document.py`

File: metagpt/roles/role.py [hunk 3]
```diff
@@ -161,26 +188,67 @@ def _set_react_mode(self, react_mode: str, max_react_loop: int = 1):
...
-    def set_env(self, env: 'Environment'):
-        """Set the environment in which the role works. The role can talk to the environment and can also receive messages by observing."""
+    def set_env(self, env: "Environment"):
+        """Set the environment in which the role works. The role can talk to the environment and can also receive
+        messages by observing."""
         self._rc.env = env
+        if env:
+            env.set_subscription(self, self._subscription)
+
+    # # Replaced by FileRepository.set_file
```
Comment (better629) [inline metagpt/roles/role.py:221]: below seems not related to Role, can be delete

File: metagpt/schema.py [hunk 4]
```diff
@@ -4,32 +4,145 @@
...
 
 
 class RawMessage(TypedDict):
     content: str
     role: str
 
 
-@dataclass
-class Message:
+class Document(BaseModel):
```
Comment (better629) [inline metagpt/schema.py:46]: what's the difference with `Document` in metagpt/document.py and `Document` in metagpt/document_store/document.py

File: metagpt/actions/action.py [hunk 5]
```diff
@@ -26,16 +26,25 @@ def __init__(self, name: str = "", context=None, llm: LLM = None):
...
+        # self.env = None
+
+    # def set_env(self, env):
+    #     self.env = env
 
     def set_prefix(self, prefix, profile):
         """Set prefix for later usage"""
         self.prefix = prefix
         self.profile = profile
+        self.llm.system_prompt = prefix
```
[the remark this answers is not in the document]
Comment (geekan) [reply metagpt/actions/action.py:46]: Now system_prompt will be modified directly, and it will be done after the role obtains llm. This logic is reasonable, and this line can be deleted (because it is a duplicate).

File: metagpt/actions/action_node.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        return {self.key: (self.expected_type, ...)}
+
+    def get_mapping(self, mode="children") -> Dict[str, Type]:
+        """get key: type mapping under mode"""
+        if mode == "children" or (mode == "auto" and self.children):
+            return self.get_children_mapping()
+        return self.get_self_mapping()
+
+    @classmethod
+    def create_model_class(cls, class_name: str, mapping: Dict[str, Type]):
```
[the remark this answers is not in the document]
Comment (geekan) [reply metagpt/actions/action_node.py:120]: This will be optimized later

File: metagpt/team.py [hunk 7]
```diff
@@ -30,33 +34,38 @@ class Config:
...
-        logger.info(self.json())
+        logger.info(self.json(ensure_ascii=False))
 
     async def run(self, n_round=3):
         """Run company until target round or no money"""
         while n_round > 0:
             # self._save()
             n_round -= 1
-            logger.debug(f"{n_round=}")
+            logger.debug(f"max {n_round=} left.")
```
Comment (better629) [inline metagpt/team.py:66]: suggest `logger.debug(f"max {n_round=} left.")` before `n_round -= 1`, or it will print max n_round=0 left, but still to run the last round. It will make a mistake.

File: metagpt/document.py (hunk 2 above)
Comment (iorisa) [reply metagpt/document.py:27]: The two code files have different functionalities, and this has been annotated for clarification. metagpt/document_store/document.py is still under design; please ignore it for now.

File: metagpt/schema.py (hunk 4 above)
Comment (iorisa) [reply metagpt/schema.py:46]: - `Document` in `shema.py` is a data structure in memory used to point to files in the file system.
- `metagpt/document.py` represents files in the file system.
- `metagpt/document_store/document.py` represents files in the vector database.

File: metagpt/team.py (hunk 7 above)
Comment (iorisa) [reply metagpt/team.py:66]: "xx left" is correct

**Comment by geekan** [comment]
LGTM

File: metagpt/schema.py (hunk 4 above)
Comment (better629) [reply metagpt/schema.py:46]: maybe with different class name is more clear.

File: metagpt/team.py (hunk 7 above)
Comment (better629) [reply metagpt/team.py:66]: what i mean is that n_round=0 means it won't run more round, but there still a last round to run now.

**Comment by kkdev163** [comment]
@garylin2099  maybe we can replace the link of RFC116、RFC135  with [RFC116](https://docs.deepwisdom.ai/main/en/rfcs/RFC-116-MetaGPT%E4%BC%98%E5%8C%96%E6%96%B9%E6%A1%88.html) and [RFC135](https://docs.deepwisdom.ai/main/en/rfcs/RFC-135-MetaGPT%20Software%20Company%E5%A2%9E%E9%87%8F%E9%9C%80%E6%B1%82%E5%BC%80%E5%8F%91%E6%96%B9%E6%A1%88%E8%AE%BE%E8%AE%A1.html), which can access by public.
